### PR TITLE
New features and code refactoring

### PR DESCRIPTION
**Pull request by sunqm**
Add new features:
* multi-collinear functional for PBC DFT
* non-local functional (VV10) for PBC DFT
* "undo" method for dynamic classes. This method can revert the action for conversion methods such as "density_fit()", "x2c()", "newton()", "as_scanner()", etc.
* Merged basis parser for molecular GTO basis and GTH basis. GTH basis can be assigned to Mole object directly.
* Merged ECP  and PP parser. PP can be assigned to Mole object directly.

Refactor:
* The layout of the Mole class and Cell class. Remove the inheritance between Cell and Mole.
* The layout of various SCF classes and the conversion methods (to_ks, to_hf, to_uhf, to_ghf, etc.) between different SCF objects.
* The layout of various MCSCF classes.
* New style to generate dynamic class for DF methods, X2C, QMMM, SOSCF, Solvent model, StateAverageMCSCF, Scanner methods, etc.
* SCF smearing method.
* Make Mole and Cell object picklable (issue #1801, #1806)
* Fix various deprecated warnings from numpy

### CODE REVIEW

**Comment by verena-neufeld** [comment]
I have only looked through parts of this. Thanks so much Qiming! The CC refactoring with a base class is a good idea. Do you think this is also a good idea for MP2 (if I missed this, please let me know)? When you add commented out tests, do you intend for them to stay commented? Again, thanks!

**Comment by sunqm** [comment]
> I have only looked through parts of this. Thanks so much Qiming! The CC refactoring with a base class is a good idea. Do you think this is also a good idea for MP2 (if I missed this, please let me know)? When you add commented out tests, do you intend for them to stay commented? Again, thanks!

Sure. The existing class hierarchy may not be convenient enough for the complex features. Please feel free to refactor them, like the MP2 class. I didn't update throughout all modules. I only updated the modules which seems too messy.

Which tests do you mean that was commented out? I moved some code under `__name__ == '__main__'` to unittests so that they will be run by the pytest pipeline.

File: pyscf/dft/test/test_r_numint.py [hunk 1]
```diff
@@ -116,7 +116,75 @@ def test_vxc1(self):
         res = mf._numint.get_vxc(mol, mf.grids, mf.xc, dm, spin=0)
         self.assertAlmostEqual(res[1], -8.631807003163278, 11)
 
-#    def test_fxc(self):
```
Comment (verena-neufeld) [inline pyscf/dft/test/test_r_numint.py:119]: This is a file I meant that has commented out tests

**Comment by verena-neufeld** [comment]
Thanks @sunqm , that makes sense! I have commented in a file that has commented out tests.

**Comment by verena-neufeld** [comment]
This PR might help issue #1737

**Review by fishjojo — APPROVED** [review]

**Pull request merged**